File: leet-code/17. Letter Combinations of a Phone Number/LetterCombinationsOfAPhoneNumber17_2.java
package com.example;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class LetterCombinationsOfAPhoneNumber17_2 {
    private final Map<String, List<String>> map = new HashMap<>(){{
        put("2", List.of("a", "b", "c"));
        put("3", List.of("d", "e", "f"));
        put("4", List.of("g", "h", "i"));
        put("5", List.of("j", "k", "l"));
        put("6", List.of("m", "n", "o"));
        put("7", List.of("p", "q", "r", "s"));
        put("8", List.of("t", "u", "v"));
        put("9", List.of("w", "x", "y", "z"));
    }};

    public List<String> letterCombinations(String digits) {
        return combinations(digits, new ArrayList<>(), new StringBuilder());
    }

    private List<String> combinations(String digits, List<String> combs, StringBuilder comb) {
        if (digits.isEmpty()) {
            if (!comb.isEmpty()) {
                combs.add(comb.toString());
            }
            return combs;
        }

        String digit = digits.substring(0, 1);
        for (String value : map.get(digit)) {
//            comb.append(value);
//            combinations(digits.substring(1), combs, comb);
//            comb.deleteCharAt(comb.length() - 1);
//            위의 코드를 아래처럼 한줄로 구현할 수 있다
            combinations(digits.substring(1), combs, new StringBuilder(comb).append(value));
        }
        return combs;
    }
}

File: leet-code/17. Letter Combinations of a Phone Number/LetterCombinationsOfAPhoneNumber17_2.kt
package com.example

class LetterCombinationsOfAPhoneNumber17_2 {
    private val map = hashMapOf(
        '2' to listOf('a', 'b', 'c'),
        '3' to listOf('d', 'e', 'f'),
        '4' to listOf('g', 'h', 'i'),
        '5' to listOf('j', 'k', 'l'),
        '6' to listOf('m', 'n', 'o'),
        '7' to listOf('p', 'q', 'r', 's'),
        '8' to listOf('t', 'u', 'v'),
        '9' to listOf('w', 'x', 'y', 'z'),
    )

    fun letterCombinations(digits: String): List<String> {
        val answer = mutableListOf<String>()
        recursion(digits, answer, StringBuilder())
        return answer
    }

    private fun recursion(digits: String, combinations: MutableList<String>, combination: StringBuilder) {
        if (digits.isEmpty()) {
            combinations.add(combination.toString())
            return
        }
        for (c in map[digits[0]]!!) {
            combination.append(c)
            recursion(digits.substring(1), combinations, combination)
            combination.deleteCharAt(combination.length - 1)
        }
    }
}
